ion/integration_test.go

```diff
@@ -0,0 +1,809 @@
+	if r.Err() != nil {
+		return r.Err()
+	}
+
+	return nil
+}
+
+func testEquivalency(t *testing.T, fp string, eq bool) {
+	file, er := os.Open(fp)
+	if er != nil {
+		t.Fatal(er)
+	}
+	defer file.Close()
+
+	r := NewReader(file)
+	for r.Next() {
+		embDoc := embeddedDoc(r.Annotations())
+		switch r.Type() {
+		case StructType, ListType, SexpType:
+			var values []item
+			r.StepIn()
+			if embDoc {
+				values = handleEmbDoc(t, r)
+			} else {
+				for r.Next() {
+					values = append(values, eqv(t, r))
+				}
+			}
+			equivalencyAssertion(t, values, eq)
+			r.StepOut()
+		}
+	}
+	if r.Err() != nil {
+		t.Error()
+	}
+}
+
+func handleEmbDoc(t *testing.T, r Reader) []item {
+	var values []item
+	for r.Next() {
+		str, err := r.StringValue()
+		if err != nil {
+			t.Error("Must be string value.")
+		}
+		newReader := NewReaderStr(str)
+		for newReader.Next() {
+			values = append(values, eqv(t, newReader))
+		}
+	}
+	return values
+}
+
+func embeddedDoc(an []string) bool {
+	for _, a := range an {
+		if a == "embedded_documents" {
+			return true
+		}
+	}
+	return false
+}
+
+func equivalencyAssertion(t *testing.T, values []item, eq bool) {
+	// Nested for loops to evaluate each item with all the other items in the list/struct/sexp
+	for i := 0; i < len(values); i++ {
+		for j := i + 1; j < len(values); j++ {
+			if i == j {
+				continue
+			}
+			if eq {
+				if !reflect.DeepEqual(values[i].value, values[j].value) ||
+					!reflect.DeepEqual(values[i].annotations, values[j].annotations) ||
+					!reflect.DeepEqual(values[i].ionType, values[j].ionType) {
+					t.Error("Equivalency test failed. All values should interpret equal.")
+				}
+			} else {
+				if reflect.DeepEqual(values[i].value, values[j].value) &&
+					reflect.DeepEqual(values[i].annotations, values[j].annotations) &&
+					reflect.DeepEqual(values[i].ionType, values[j].ionType) {
+					t.Error("Non-Equivalency test failed. Values should not interpret equal.")
+				}
+			}
+		}
+	}
+}
+
+func readFilesAndTest(t *testing.T, path string, skipList []string, tf testingFunc) {
+	files, err := ioutil.ReadDir(path)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if debugFile != "" {
+		fp := filepath.Join(path, debugFile)
+		t.Run(fp, func(t *testing.T) {
+			tf(t, fp)
+		})
+	} else {
+		for _, file := range files {
+			fp := filepath.Join(path, file.Name())
+			if file.IsDir() {
+				readFilesAndTest(t, fp, skipList, tf)
+			} else if skipFile(skipList, file.Name()) {
+				continue
+			} else {
+				t.Run(fp, func(t *testing.T) {
+					tf(t, fp)
+				})
+			}
+		}
+	}
+}
+
+func loadFile(t *testing.T, path string) []byte {
+	b, err := ioutil.ReadFile(path)
+	if err != nil {
+		t.Fatal(err)
+	}
+	return b
+}
+
+func skipFile(skipList []string, fn string) bool {
+	ion, _ := regexp.MatchString(`.ion$`, fn)
+	bin, _ := regexp.MatchString(`.10n$`, fn)
+
+	return !ion && !bin || isInSkipList(skipList, fn)
+}
+
+func isInSkipList(skipList []string, fn string) bool {
+	for _, a := range skipList {
```

Comment:
Since there are hundreds of files and the skip lists are pretty long, scanning the entire skip list for each file could get a bit slow. Can we make `skipList` a `Set` of some kind?

Reply:
AFAIK Go does not have ```Set```. The closest thing would be introducing a ```map``` and set filenames as the key. This is only good to make sure we will not have duplicates, but as I was reading there is not a big difference between traversing map versus slice. Seems like slice has a better look up time when the the size is smaller 100 and map can do better for bigger numbers.
Let me know what you think.

Reply:
Cool, thanks.